pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,1323 @@
+		///
+		/// Parameters:
+		/// - `origin`:
+		/// - `total_rewards`:
+		/// - `planned_yielding_periods`: planned number of periods to distribute rewards. WARN: this is not
+		/// how long will farming run.  Owner can destroy farm sooner or liq. mining can run longer
+		/// if all the rewards will not distributed.
+		/// - `blocks_per_period`:
+		/// - `incentivized_token`
+		/// - `reward_currency`
+		/// - `admin_account`
+		/// - `yield_per_period`
+		#[pallet::weight(1000)]
+		#[transactional]
+		pub fn create_farm(
+			origin: OriginFor<T>,
+			total_rewards: Balance,
+			planned_yielding_periods: PeriodOf<T>,
+			blocks_per_period: BlockNumberFor<T>,
+			incentivized_token: AssetIdOf<T>,
+			reward_currency: AssetIdOf<T>,
+			owner: AccountIdOf<T>,
+			yield_per_period: Permill,
+		) -> DispatchResult {
+			T::CreateOrigin::ensure_origin(origin)?;
+
+			Self::validate_create_farm_data(
+				total_rewards,
+				planned_yielding_periods,
+				blocks_per_period,
+				yield_per_period,
+			)?;
+
+			ensure!(
+				T::MultiCurrency::free_balance(reward_currency, &owner) >= total_rewards,
+				Error::<T>::InsufficientRewardCurrencyBalance
+			);
+
+			let planned_periods =
+				Balance::from(TryInto::<u128>::try_into(planned_yielding_periods).map_err(|_e| Error::<T>::Overflow)?);
+			let max_reward_per_period = total_rewards.checked_div(planned_periods).ok_or(Error::<T>::Overflow)?;
+			let now_period = Self::get_now_period(blocks_per_period)?;
+			let pool_id = Self::get_next_id()?;
+
+			let pool = GlobalPool::new(
+				pool_id,
+				now_period,
+				reward_currency,
+				yield_per_period,
+				planned_yielding_periods,
+				blocks_per_period,
+				owner,
+				incentivized_token,
+				max_reward_per_period,
+			);
+
+			<GlobalPoolData<T>>::insert(&pool.id, &pool);
+
+			let pool_account = Self::pool_account_id(pool.id)?;
+			T::MultiCurrency::transfer(reward_currency, &pool.owner, &pool_account, total_rewards)?;
+
+			Self::deposit_event(Event::FarmCreated(pool.id, pool));
+
+			Ok(())
+		}
+
+		#[pallet::weight(1000)]
+		pub fn destroy_farm(origin: OriginFor<T>, farm_id: PoolId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			<GlobalPoolData<T>>::try_mutate_exists(farm_id, |maybe_g_pool| -> DispatchResult {
+				let g_pool = maybe_g_pool.as_ref().ok_or(Error::<T>::FarmNotFound)?;
+
+				ensure!(who == g_pool.owner, Error::<T>::Forbidden);
+
+				ensure!(g_pool.liq_pools_count.is_zero(), Error::<T>::FarmIsNotEmpty);
+
+				let g_pool_account = Self::pool_account_id(g_pool.id)?;
+				//NOTE: should this be 0?
+				ensure!(
+					T::MultiCurrency::free_balance(g_pool.reward_currency, &g_pool_account).is_zero(),
+					Error::<T>::RewardBalanceIsNotZero
+				);
+
+				*maybe_g_pool = None;
+
+				Self::deposit_event(Event::FarmDestroyed(farm_id, who));
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(1000)]
+		pub fn withdraw_undistributed_rewards(_origin: OriginFor<T>, _farm_id: PoolId) -> DispatchResult {
+			todo!()
```

Comment:
Also this todo. Note that if you call this functions , the todo macro calls panic.